core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+        """Update counting matrices given doc/word/label indices.
+        Either increments or decrements."""
+        val = 1 if inc else -1
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._num_docs)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dp.append(l_c)
+            for w, l in zip(doc, l_c):
+                self._update_counting_matrices(d, w, l)
+
+    def _infer_docs(self, doc_ids):
+        """Runs iterative inference on given docs."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for p, w in enumerate(doc):
+                l = self._l_dp[d][p]
+
+                self._update_counting_matrices(d, w, l, False)
+
+                # Gibbs update of labels
+                coeff_a = 1.0 / (
+                    self._c_dl[d].sum() + self._num_labels * self._alpha)
+                coeff_b = 1.0 / (
+                    self._c_lw.sum(axis=1) + self._num_words * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
+                    1, prob_l / prob_l.sum()).argmax()
+
+                statez['computes'] += 1
+                if l != new_l:
+                    statez['updates'] += 1
+
+                self._l_dp[d][p] = new_l
+                self._update_counting_matrices(d, w, new_l)
+
+        return statez
+
+    def _get_probability(self, d):
+        """Returns the probability of a doc having a label."""
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probability_with_label(self, d):
+        """Returns a dict that maps human-readable labels to the probability
+        that the doc has the label."""
+        probs_i = self._get_probability(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc.
+        Returns a tuple of (predicted label, prediction score)."""
+        probs_l = self._get_probability_with_label(d)
```

Comment:
This is the part where all the internal magic is resurfaced back to the client, and the client has no idea labels have ids. So the ids need to be re-mapped to what was originally passed in.